torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2003,6 +2012,245 @@ def clip_grad_norm_(
                 assert p.grad is not None
                 p.grad.detach().mul_(clip_coef.to(p.grad.device))
 
+    @staticmethod
+    def full_optim_state_dict(
+        model: torch.nn.Module,
+        optim: torch.optim.Optimizer,
+        optim_input: Optional[Union[
+            List[Dict[str, Any]], Iterable[torch.nn.Parameter],
+        ]] = None,
+    ) -> Dict[str, Any]:
+        """
+        Consolidates the full optimizer state on the target rank and returns it
+        as a :class:`dict` following the convention of normal
+        :class:`torch.optim.Optimizer` s. This involves remapping from
+        flattened to unflattened parameters for any FSDP modules contained in
+        ``model`` so that the state is agnostic to the current world size and
+        resulting sharding scheme.
+
+        The returned :class:`dict` is identical to the result of
+        ``torch.optim.Optimizer.state_dict()`` for an optimizer constructed
+        using the parameters of a purely non-wrapped version of the model (i.e.
+        without any ``FullyShardedDataParallel`` modules).
+
+        .. warning:: This needs to be called on all ranks since synchronization
+            primitives are used.
+
+        .. note:: Tensors contained in the optimizer state dict are not
+        cloned, so there may be aliasing surprises. For best practices, save
+        the optimizer state dict to disk immediately (e.g. using
+        ``torch.save()``).
+
+        Args:
+            model (torch.nn.Module): Root module (which may or may not be a
+                :class:`FullyShardedDataParallel` instance) whose parameters
+                were passed into the optimizer ``optim``.
+            optim (torch.optim.Optimizer): Optimizer for ``model`` 's
+                parameters.
+            optim_input (Optional[Union[List[Dict[str, Any]],
+            Iterable[torch.nn.Parameter]]]): Input passed into the optimizer
+                ``optim`` representing either a :class:`list` of parameter
+                groups or an iterable of parameters; if ``None``, then this
+                method assumes the input was ``model.parameters()``. (Default:
+                ``None``)
+
+        Returns:
+            full_osd (Dict[str, Any]): A :class:`dict` containing the optimizer
+                state for ``model`` 's original unflattened parameters and
+                having keys "state" and "param_groups" following the convention
+                of normal :class:`torch.optim.Optimizer` s.
+        """
```

Comment:
To-do: Add warning about not supporting changing the `FSDP` wrapping scheme.

Reply:
To-do: Store parameter ID mapping in the optim state dict to error early (as is, my implementation does not error early consistently).